File: binary_search/binsearch/c/binsearch.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int binsearch( int arr[], int low, int high, int mem){
    if( high <low){
        return -1;    
    }

    int mid = low + ((high-low)>>1);
    if( mem == arr[mid]){
        return mid;
    }
    if(mem > arr[mid]){
        return binsearch(arr, mid+1,high, mem);
    }else{
        return binsearch(arr, low, mid-1, mem);    
    }
}

int binsearch_no(int arr[], int n, int mem){
    int low = 0; 
    int high = n-1;

    while( low <= high){
        int mid = low + ((high-low)>>1); 
        if( arr[mid] == mem){
            return mid;    
        }
        if( arr[mid]> mem){
            high = mid-1;    
        }else{
            low = mid+1;
        }
    }

    return -1;
}

int main(){

    int arr[]={1,7,9, 11, 23, 26,33,35,37,47,59};
    int n = sizeof(arr)/sizeof(arr[0]);

    int r = binsearch(arr, 0, n-1, 23);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 47);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 26);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 27);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 2);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 1);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch(arr, 0, n-1, 59);
    printf("find: %d %d\n", r , arr[r]);

    printf("---------------------------------\n");
    r = binsearch_no(arr,  n, 23);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr, n, 47);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr,  n, 26);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr,  n, 27);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr,  n, 2);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr,  n, 1);
    printf("find: %d %d\n", r , arr[r]);
    r = binsearch_no(arr,  n, 59);
    printf("find: %d %d\n", r , arr[r]);
    return 0;    
}

File: binary_search/binsearch/go/binsearch.go
package main

import (
    "fmt"
)

func binsearch(arr []int, low, high, mem int) int{
    if  low > high {
        return -1
    }

    mid := low + ((high-low)>>1)

    if mem == arr[mid]{
        return mid
    }

    if( mem < arr[mid]){
        return binsearch(arr, low, mid-1, mem)
    }else{
        return binsearch(arr, mid+1, high, mem)
    }
}

func binsearch_no(arr []int, n, mem int) int{
    low:=0
    high:=n-1

    for {
        if( low > high){
            break
        }
        mid:= low+((high-low)>>1)
        if( mem == arr[mid]){
            return mid;
        }

        if( mem < arr[mid]){
            high = mid-1
        }else{
            low = mid+1
        }
    }
    return -1 
}
func main(){
    var arr=[]int{1,3,5,7,9,11,23,25, 27,30,47,55}
    r := binsearch(arr, 0, len(arr), 11)
    fmt.Println("find: ", r, arr[r])
    r = binsearch_no(arr,  len(arr), 11)
    fmt.Println("find: ", r, arr[r])

}
